### ISSUE TITLE
Unable to update products

### ISSUE THREAD

**Issue by DannyRoberts95** [ndimatteo/HULL#68]
Hey,

First of all the project is amazing! Hella props!

I have followed the automatic setup guide and have the project up and deployed and running locally with no hassle.
The only problem I can't seem to get around is triggering the product updates from my Shopify setup to be reflected in my project data.

I have setup the env with the shopify keys, permissions as well as the web-hooks inside Shopify. 
I have also updated products to try and trigger the update but so far no dice.

How does one trigger the build of the sanity data from the Shopify products.

I'm relatively new to both shopify and sanity so forgive me if this is an obvious question!

Thanks

**Comment by ndimatteo** [comment]
Thanks @DannyRoberts95, glad you're digging the project!

A few things to check first:
1. For the Shopify webhooks, did you enter a URL of a deployed version of your project? You'll need to ensure the webhooks are using live URLs (no localhost) and that the deployed version also has the ENV variables in place.
2. Can you confirm if the products have the headless private app _enabled_ as a sales channel?

If that all checks out, you'll need to trigger the webhook by making a change (of virtually any kind) to each of your products within Shopify.

Once you do that, the webhook should trigger and send the data to the live URL supplied there. If everything is setup properly it should pull that data into your Sanity Project dataset.

If you've deployed to Vercel, I'd also suggest watching the function logs for the sync to see that it's (a) firing, and (b) how it's going through the process.

Let me know if that helps!

**Comment by DannyRoberts95** [comment]
Hey @ndimatteo, thanks for the advice. Still having some issues but moving in the right direction.

Yup, I have added the headless app to the products sales channels (nice catch). I'm able to make changes and fire off test webhooks from shopify which can be seen inside of my Vercel logs so at least I know they are talking to each other.  

Currently logs are spitting back a 404 error upon firing update/delete. Look like it's looking to update/ delete a sanity document that doesn't exist? Is document **7501112901882** perhaps a doc in the Hull pre populated documents?

This makes me think it's perhaps a config issue on my end? I've now spent a few hours confirming each env var is correct.

I'll attach the current env and err (with slightly augmented values) below to see if anything jumps out as incorrect.

Many thanks my dude!



** Webhook function err

```
[POST] /api/shopify/product-delete
15:59:40:66
Function Status:
None
Edge Status:
500
Duration:
426.90 ms
Init Duration:
442.47 ms
Memory Used:
84 MB
ID:
442zn-1640707180657-44e7a30302bb
User Agent:
Shopify-Captain-Hook
'50',
      'ratelimit-remaining': '49',
      'x-sanity-shard': 'gcp-eu-w1-01-prod-1021',
      'x-served-by': 'gradient-web-78bd6d78dd-n7z4j',
      date: 'Tue, 28 Dec 2021 15:59:41 GMT',
      vary: 'Origin',
      xkey: 'project-ssynjzos, project-ssynjzos-production',
      via: '1.1 google',
      'alt-svc': 'clear',
      connection: 'close'
    },
    statusCode: 404,
    statusMessage: 'Not Found'
  },
  statusCode: 404,
  responseBody: '{\n' +
    '  "error": {\n' +
    '    "description": "the mutation(s) failed: The document with the ID \\"product-7501112901882\\" was not found",\n' +
    '    "items": [\n' +
    '      {\n' +
    '        "error": {\n' +
    '          "description": "The document with the ID \\"product-7501112901882\\" was not found",\n' +
    '          "id": "product-7501112901882",\n' +
    '          "type": "documentNotFoundError"\n' +
    '        },\n' +
    '        "index": 0\n' +
    '      }\n' +
    '    ],\n' +
    '    "type": "mutationError"\n' +
    '  }\n' +
    '}',
  details: {
    description: 'the mutation(s) failed: The document with the ID "product-7501112901882" was not found',
    items: [ [Object] ],
    type: 'mutationError'
  }
}
2021-12-28T15:59:41.823Z	214921e1-a648-48ba-b228-6e1f40ef12fa	ERROR	ClientError: the mutation(s) failed: The document with the ID "product-7501112901882" was not found
    at onResponse (/var/task/node_modules/@sanity/client/lib/http/request.js:27:13)
    at applyMiddleware (/var/task/node_modules/get-it/lib-node/util/middlewareReducer.js:14:23)
    at onResponse (/var/task/node_modules/get-it/lib-node/index.js:82:22)
    at /var/task/node_modules/get-it/lib-node/index.js:49:55
    at callback (/var/task/node_modules/get-it/lib-node/request/node-request.js:57:46)
    at /var/task/node_modules/get-it/lib-node/request/node-request.js:147:14
    at DestroyableTransform.<anonymous> (/var/task/node_modules/simple-concat/index.js:8:13)
    at Object.onceWrapper (events.js:519:28)
    at DestroyableTransform.emit (events.js:412:35)
    at endReadableNT (/var/task/node_modules/readable-stream/lib/_stream_readable.js:1010:12) {
  response: {
    body: { error: [Object] },
    url: 'https://ddynjzos.api.sanity.io/v2021-03-25/data/mutate/production?returnIds=true&visibility=sync',
    method: 'POST',
    headers: {
      'content-type': 'application/json; charset=utf-8',
      'content-length': '313',
      'ratelimit-reset': '1',
      'x-ratelimit-remaining-second': '49',
      'x-ratelimit-limit-second': '50',
      'ratelimit-limit': '50',
      'ratelimit-remaining': '49',
      'x-sanity-shard': 'gcp-eu-w1-01-prod-1021',
      'x-served-by': 'gradient-web-78bd6d78dd-n7z4j',
      date: 'Tue, 28 Dec 2021 15:59:41 GMT',
      vary: 'Origin',
      xkey: 'project-ssynjzos, project-ssynjzos-production',
      via: '1.1 google',
      'alt-svc': 'clear',
      connection: 'close'
    },
    statusCode: 404,
    statusMessage: 'Not Found'
  },
  statusCode: 404,
  responseBody: '{\n' +
    '  "error": {\n' +
    '    "description": "the mutation(s) failed: The document with the ID \\"product-7501112901882\\" was not found",\n' +
    '    "items": [\n' +
    '      {\n' +
    '        "error": {\n' +
    '          "description": "The document with the ID \\"product-7501112901882\\" was not found",\n' +
    '          "id": "product-7501112901882",\n' +
    '          "type": "documentNotFoundError"\n' +
    '        },\n' +
    '        "index": 0\n' +
    '      }\n' +
    '    ],\n' +
    '    "type": "mutationError"\n' +
    '  }\n' +
    '}',
  details: {
    description: 'the mutation(s) failed: The document with the ID "product-7501112901882" was not found',
    items: [ [Object] ],
    type: 'mutationError'
  }
}
RequestId: 214921e1-a648-48ba-b228-6e1f40ef12fa Error: Runtime exited with error: exit status 1
Runtime.ExitError
```

** ENV example (not my actualy values)
```

SHOPIFY_STORE_ID = my-store-name
SHOPIFY_API_TOKEN = 1234a36b48c6d727e234f65e4336f74e
SHOPIFY_API_PASSWORD = plzwrk_1e6515f814fd34d14cd0as3104562nin
SHOPIFY_WEBHOOK_INTEGRITY = b12g34a956f89562199c2567e66da4a1q32b4450d141480a50f82381c2225a99

SANITY_PROJECT_DATASET=production
SANITY_PROJECT_ID=1ijshlre
SANITY_API_TOKEN=drPqpythZqfNqtEoeYvpm4vF3fOtJjkR8cjuWLBkl1EdANVRA4MTuoI5dTmtwoyTCKSwDBbf3N7MOqpEXfAsQPA889YrETMrCYPfR5oWttGOa6cstjdAvj0qzge0amNFTI0ln6gb2RbXvvDuBLqZnTnRqPulPHziP3Ki5PbZJW4aLKKf0HH2

...other vars are blank, havn't set em up yet

```

**Comment by ndimatteo** [comment]
thanks @DannyRoberts95!

It sounds like the Sanity API update is running into an issue with your dataset.

I'm assuming that this error is coming from this part of the sync (where we create or mutate the product document): https://github.com/ndimatteo/HULL/blob/main/pages/api/shopify/product-update.js#L246

Here's info on that method from the Sanity client docs: https://www.sanity.io/docs/js-client#creating-if-not-already-apresent

We're using that in conjunction with the `transaction()` method to chain what we're doing. That sequence is this:
1. `create` a document, unless that document already exists
2. `patch` the new/existing document with our product info
3. `commit` all of that to your Sanity dataset

The error you're getting doesn't make a ton of sense, since that first `createIfNotExists` piece is meant to handle just that. So if it wasn't created, that means it exists, and to simply continue with patching the existing document.

I'd take a closer look at your Sanity project settings and associated ENV variables. Notably the `SANITY_API_TOKEN` variable: make sure you've given that read _**and**_ write access.

If you're still running into issues, it might be worth posting this in the Sanity community slack as well, to get some official Sanity eyes on it!

I hope that helps, and let me know if you get unblocked 🤘

**Comment by ndimatteo** [comment]
@DannyRoberts95 please start a new issue if this is persisting and you have additional info. Thanks!
